### PR TITLE
G-API: Introduce streaming::desync and infer(ROI)

### PR DESCRIPTION
**Pull request by dmatveev**
This patch is mainly upstreaming of a long-live internal branch.
The functionality has been tested already in the "real world" scenarios.

- desync() is a new (and for now, the only one) intrinsic
  which splits the graph execution into asynchronous parts
  when running in Streaming mode;
- desync() makes no effect when compiling in Traditional mode;
- Added tests on desync() working in various scenarios;
- Extended GStreamingExecutor to support desync(); also extended
  GStreamingCompiled() with a new version of pull() returning a
  vector of optional values;
- Fixed various issues with storing the type information & proper
  construction callbacks for GArray<> and GOpaque;

- Introduced a new infer(Roi,GMat) overload with a sample;

- Introduced an internal API for Islands to control fusion
  procedure (to fuse or not to fuse);
- Introduced handleStopStream() callback for island executables;
- Added GCompileArgs to metadata of the graph (required for other
  features).

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom,Custom Win,Custom Mac
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
build_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f

Xbuild_image:Custom=centos:7
Xbuildworker:Custom=linux-1
build_gapi_standalone:Custom=ade-0.1.1f

build_image:Custom=ubuntu-openvino-2020.3.0:16.04
build_image:Custom Win=openvino-2020.3.0
build_image:Custom Mac=openvino-2020.3.0

test_modules:Custom=gapi
test_modules:Custom Win=gapi
test_modules:Custom Mac=gapi

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

**Comment by dmatveev** [comment]
@TolyaTalamanov @rgarnov builds are green. 

Let's have a look on this one more time and get it merged then

**Review by rgarnov — APPROVED** [review]

File: modules/gapi/test/streaming/gapi_streaming_tests.cpp [hunk 1]
```diff
@@ -1018,4 +1129,380 @@ TEST(Streaming, Python_Pull_Overload)
...
+    std::size_t out2_hits = 0u;
+    cv::optional<cv::Mat> test_out1, test_out2;
+    while (sc.pull(cv::gout(test_out1, test_out2))) {
+        GAPI_Assert(test_out1 || test_out2);
+        if (test_out1) out1_hits++;
+        if (test_out2) out2_hits++;
+    }
+    EXPECT_EQ(100u, out1_hits);      // out1 must be available for all frames
+    EXPECT_LE(out2_hits, out1_hits); // out2 must appear less times than out1
+    std::cout << "Got " << out1_hits << " out1's and " << out2_hits << " out2's" << std::endl;
```
[the remark this answers is not in the document]
Comment (dmatveev) [reply modules/gapi/test/streaming/gapi_streaming_tests.cpp:1176]: This is mainly to visually inspect out1 vs out2

File: modules/gapi/test/streaming/gapi_streaming_tests.cpp [hunk 2]
```diff
@@ -1018,4 +1129,380 @@ TEST(Streaming, Python_Pull_Overload)
...
+    // Desynchronized path 2 (depends on 1)
+    cv::GMat tmp3 = cv::gapi::streaming::desync(tmp1);
+    cv::GMat out2 = 0.5*tmp3 + tmp2;
+
+    // This shouldn't compile
+    EXPECT_ANY_THROW(cv::GComputation(cv::GIn(in), cv::GOut(out1, out2))
+                     .compileStreaming());
+}
+
+TEST(GAPI_Streaming_Desync, Negative_OtherDesync_Tier1)
```
[the remark this answers is not in the document]
Comment (dmatveev) [reply modules/gapi/test/streaming/gapi_streaming_tests.cpp:1295]: you mean, when there's a link between one desync path and another?

[the remark this answers is not in the document]
Comment (rgarnov) [reply modules/gapi/test/streaming/gapi_streaming_tests.cpp:1295]: Just a naming, previous test has a `Cross` prefix

**Review by TolyaTalamanov — APPROVED** [review]

File: modules/gapi/test/streaming/gapi_streaming_tests.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (dmatveev) [reply modules/gapi/test/streaming/gapi_streaming_tests.cpp:1295]: Got it. Fixed, thanks!

**Comment by dmatveev** [comment]
@alalek @mshabunin review complete, required builds passed. Can you please merge it? We need to have this in our baseline before next Wed. Thanks!

File: modules/gapi/samples/infer_single_roi.cpp [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+std::string weights_path(const std::string &model_path) {
+    const auto EXT_LEN = 4u;
+    const auto sz = model_path.size();
+    CV_Assert(sz > EXT_LEN);
+
+    auto ext = model_path.substr(sz - EXT_LEN);
+    std::transform(ext.begin(), ext.end(), ext.begin(), [](unsigned char c){
+            return static_cast<unsigned char>(std::tolower(c));
+        });
```
Comment (dmatveev) [inline modules/gapi/samples/infer_single_roi.cpp:34]: so I know about cv::toLowerCase but let's address this in another MR (with a test on InferROI)

**Comment by dmatveev** [comment]
Fantastic! Thanks!

**Pull request merged**